From d3e058db893910a0faf72f3f652c35a5694d3b1b Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@kernel.org>
Date: Tue, 4 Apr 2023 11:59:00 -0400
Subject: [PATCH 312/780] [Backport] dm clone: call kmem_cache_destroy() in
 dm_clone_init() error path

stable inclusion
from stable-v5.10.180
commit 717ceb487bc30bc32ccb99fa5966a4f5f4e84686
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I7FF4R

Reference: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?id=717ceb487bc30bc32ccb99fa5966a4f5f4e84686

--------------------------------

commit 6827af4a9a9f5bb664c42abf7c11af4978d72201 upstream.

Otherwise the _hydration_cache will leak if dm_register_target() fails.

Cc: stable@vger.kernel.org
Signed-off-by: Mike Snitzer <snitzer@kernel.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Wang Hai <wanghai38@huawei.com>
---
 drivers/md/dm-clone-target.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/md/dm-clone-target.c b/drivers/md/dm-clone-target.c
index 0d38ad623534..e3156b30294f 100644
--- a/drivers/md/dm-clone-target.c
+++ b/drivers/md/dm-clone-target.c
@@ -2221,6 +2221,7 @@ static int __init dm_clone_init(void)
 	r = dm_register_target(&clone_target);
 	if (r < 0) {
 		DMERR("Failed to register clone target");
+		kmem_cache_destroy(_hydration_cache);
 		return r;
 	}
 
-- 
2.17.1

